File: Data Structures & Algorithms/extra-characters-in-a-string/submission-5.py
class TrieNode:
    def __init__(self):
        self.childern={}
        self.isWord=False


class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        cache = {len(s):0}
        root = TrieNode()
        for w in dictionary:
            curr=root
            for c in w:
                if c not in curr.childern:
                    curr.childern[c] = TrieNode()
                curr=curr.childern[c]
            curr.isWord=True
        

        def dfs(i):
            if i in cache: return cache[i] 

            res = 1 + dfs(i+1)
            curr=root
            j = i

            while j<len(s) and s[j] in curr.childern:
                curr=curr.childern[s[j]]
                j+=1
                if curr.isWord:
                    res = min(res, dfs(j))
            
            cache[i] = res
            return res
        
        return dfs(0)




        

File: Data Structures & Algorithms/two-integer-sum-ii/submission-5.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int p1 = 0, p2 = numbers.size()-1;
        int curr = 0;

        while (p1<p2){
            curr = numbers[p1] + numbers[p2];

            if (curr<target) p1++;
            else if (curr>target)p2--;
            else break;
        }

        return vector{p1+1, p2+1};
        
    }
};
